### PR TITLE
Enhance R markdown support

### LINKED ISSUE #428: Enhance R Markdown support
A number of issues have been filed as requests to improve R markdown support:

* https://github.com/Ikuyadeu/vscode-R/issues/49
* https://github.com/Ikuyadeu/vscode-R/issues/261

The features could be implemented via [CodeLens provider](https://github.com/microsoft/vscode-extension-samples/tree/master/codelens-sample) and some text editor decoration (https://github.com/Ikuyadeu/vscode-R/issues/261#issuecomment-629162799).

### PR DESCRIPTION
**Pull request by renkun-ken**
**What problem did you solve?**

Close #428, #49, #261 

* [x] Run chunk and Run above command
* [x] Highlight code chunks
* [x] Respect chunk options (Run above should skip chunks with `eval=FALSE`)
* [x] Command to run current chunk
* [x] Completions for chunk options

**(If you have)Screenshot**

Kapture 2020-10-28 at 22 25 08

How to test:

1. Create a Rmd document with a number of code chunks (using `r chunk` snippet)
2. Above each chunk there are two CodeLens items: Run Chunk and Run Above.
3. Each chunk has background highlighting.
4. Click Run Chunk and the chunk code is sent to terminal
5. Click Run Above and the code above the chunk is combined and sent to terminal
6. If a chunk has option `eval=FALSE` then Run Above will not include such chunks.
7. Put cursor in a chunk and run `R: Run Current Chunk` and the chunk is sent to terminal.8
8. If cursor is outside any chunk, run `R: Run Current Chunk` and nothing happens.
9. Pressing `Ctrl+shift+enter` should trigger `R: Run Current Chunk` by default (same as RStudio).
10. Editing chunk header will trigger chunk options completion and selecting an item will insert `{label}=` to the editor.

### CODE REVIEW

**Comment by renkun-ken** [comment]
@Ikuyadeu @andycraig 

I'm wondering if we already have a shortcut/command/snippet to insert an R code chunk like the following?

````
```{r}

```
````

I tried adding a snippet to `snippets/markdown.json` but it seems that changing (even removing all content from) the file does not have any effect at all.

**Comment by MilesMcBain** [comment]
I like the background highlighting, looks great.

Would you consider an option to disable the text links? 

I like a really minimal editor layout, with things driven by keybindings. I usually program in 'zen mode'. 

Of course this is my selfish personal preference. I know many users would appreciate the links.

**Comment by renkun-ken** [comment]
>Would you consider an option to disable the text links?

By text links, do you mean the "Run Chunk" and "Run Above" CodeLens commands? You could disable CodeLens globally with the following settings:

```json
"editor.codeLens": false
```

or just disable it for Rmd:

```json
"[rmd]": {
  "editor.codeLens": false
}
```

**Comment by Ikuyadeu** [comment]
@renkun-ken 
> I tried adding a snippet to snippets/markdown.json but it seems that changing (even removing all content from) the file does not have any effect at all.
It is due to the rmd is not defined in original vscode, it will works after user installs our extensions.
After you merge this, I will try to fix it.

**Comment by renkun-ken** [comment]
@Ikuyadeu Thanks!

I've switched to our own snippets for rmd and it is much easier to insert an R chunk now.

Kapture 2020-10-29 at 22 56 33

File: snippets/rmarkdown.json [hunk 1]
```diff
@@ -67,5 +67,23 @@
 		"prefix": "image",
 		"body": "![${1:alt}](http://${2:link})$0",
 		"description": "Insert image"
+	},
+	"Insert code chunk": {
```
Comment (Ikuyadeu) [inline snippets/rmarkdown.json:71]: Duplicated with line 22 "Insert fenced code block" on the original snippet.
Also, I suggest that can follow original names such as
"Insert R code chunk" ->"Insert R fenced code block"

Comment (renkun-ken) [reply snippets/rmarkdown.json:71]: "Fenced code block" is more of a term in Markdown.
"Code chunk" is more of a term in R Markdown (as widely used in R markdown documentations e.g https://rmarkdown.rstudio.com/lesson-3.html, https://bookdown.org/yihui/rmarkdown/r-code.html) and is used in RStudio.

I suggest that we follow the "official names" of such and R markdown users might be more familiar with them?

Comment (renkun-ken) [reply snippets/rmarkdown.json:71]: From my understanding, code chunk in R Markdown specifically means something like

````
```{language}

```
````

for rmarkdown renderer to use while fenced code block means

````
```language

```
````

where `language` is not enclosed by ``{}`.

Comment (renkun-ken) [reply snippets/rmarkdown.json:71]: To be more clear, what about 

`chunk` -> `code chunk`
`r chunk` -> `r code chunk`

Comment (Ikuyadeu) [reply snippets/rmarkdown.json:71]: > I suggest that we follow the "official names" of such and R markdown users might be more familiar with them?

Thank you for your knowledge, I did not know R markdown standard names, 

your suggestion
> chunk -> code chunk
> r chunk -> r code chunk

Looks good that looks easy for both RStudio user and vs code user.

**Comment by andycraig** [comment]
@renkun-ken People (including me) have wanted these features for ages - thank you very much for this PR!

**Pull request merged**